### ISSUE TITLE
remove Deferred.setTimeout usage from twisted.mail

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2729]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2729|
|Type|task|
|Created|2007-07-07 20:46:07Z|

<details><summary>Searchable metadata</summary>

```
trac-id__2729 2729
type__task task
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1183841167000000 1183841167000000
changetime__1194451925000000 1194451925000000
version__None None
owner__ 
cc__exarkun cc__peaker cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

This is started in process-alias-timeout-2729.  I think the tests still need a bit of work, though.

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_2"></a>|@Peaker commented|
|-|-|

The branch needs to be forward-merged with trunk or restarted.

Can't a similar fix as in http://twistedmatrix.com/trac/attachment/ticket/2245/cdefer.3.patch be used? (The patch includes more things, take a look just at the alias.py fix).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

This not as simple, because your patch doesn't add tests.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

I merged this forward to **process-alias-timeout-2729-2**.  One specific area I notice is short on test coverage is ```ProcessAliasProtocol.processEnded```.  If you run ```trial --coverage twisted.mail.test```, you see that the entire ```else``` suite in that method is uncovered.  The other area of the code I'm worried about is ```MessageWrapper._processEnded```: the test suite only exercises it once - which means it only goes through the true case of the ```if self._timeoutCallID is not None:``` test, which means the conditional doesn't need to be there at all (as far as the test suite is concerned).  A test which went through this method with that conditional evaluating to false would be nice - if it is possible to do that, otherwise the check should be dropped.

I'll get to this after I finish all the reviews, but if someone beats me to it that'd be great.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Oh yea, this is blocking [#2245](https://github.com/twisted/twisted/issues/2245) which I want to review, so I guess I'll end up working on this first thing tomorrow.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

Done in **process-alias-timeout-2729-2**

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @exarkun|
|-|-|

 * unused import error in alias.py
 * ```StubProcess``` doesn't document its methods

Otherwise, that looks great, so please merge once done.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [21622]) remove unused import

refs [#2729](https://github.com/twisted/twisted/issues/2729)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [21624]) docstrings for StubProcess

refs [#2729](https://github.com/twisted/twisted/issues/2729)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [21628]) Merge process-alias-timeout-2729-2

Author: exarkun
Reviewer: therve
Fixes [#2729](https://github.com/twisted/twisted/issues/2729)

Improve test coverage for the aliases(5) process support, change the implementation
not to use Deferred.setTimeout, and fix some other cases related to process termination
which were handled incorrectly.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
